### DISCUSSION TITLE
Should we allow for functions that sample nFathers to sample 0?

### DISCUSSION THREAD

**Discussion by janaobsteter** [HighlanderLab/SIMplyBee#292] [Q&A] [answered]
In reality, if the drones are in the vicinity, the queens would fly as many times as necessary to mate. Should we then even allow for functions that pull drones groups from DCA to pull out 0 drones?

**Comment by gregorgorjanc** [comment]
I think it depends on what a user would like to model. I thinking it’s important to split: i) biology, ii) function behaviour, iii) flexibility - at least as much as possible. Obviously we should follow biology as much as possible!!! But, we can make the functions a bit more general and then give the user option to follow biology closely or deviate from it somewhat. Ideally, the default should be in line with biology.

This is why I was not happy with the Poisson distribution in sampling the number of fathers in https://github.com/HighlanderLab/SIMplyBee/issues/84 and suggested zero-truncated Poisson.

**Reply by janaobsteter** [reply]
Yes, I think we should provide the users with the truncated Poisson only (?)

**Reply by gregorgorjanc** [reply]
We provide 3 options:

1) Fixed number
2) Poisson distribution
3) Truncated Poisson distribution

It depends what one wants to simulate. Your suggestion would mean that mating is ALWAYS successful. Is that a reasonable assumption?

**Comment by janaobsteter** [comment]
If other virgin queens are mated successfully, it means there are drones around - hence the virgin queen would take as many mating flights as needed to get mated. The only option that the mating is unsuccessful is if she doesn't return (this is summarizing Janez's words).

**Comment by gregorgorjanc** [comment]
Which is why we have the truncated Poisson distribution. I don’t buy that mating is always successful though, hence good to have Poisson as an option should one need it.

**Comment by janaobsteter** [comment]
Ok, just to write down current functionality. We now have a function `cross` that caters for mating everything (`Pop`, `Colony`, `Colonies`).
If the input is:
- a single virgin queen: if mated with 0 fathers (and `checkMating = TRUE`), the function stops, returns nothing, and the virgin queen remains virgin
- a `Pop` of virgin queens: as soon as one is mated with 0 fathers, the function stops and doesn't mate the rest of them. Is this what we want? We do have a parameter `checkMating` that controls this - if set to "warning", the function continues
- `Colony` and `Colonies` resemble the behavior above


Here is the code for these tests:
```
founderGenomes <- quickHaplo(nInd = 20, nChr = 1, segSites = 100)
SP <- SimParamBee$new(founderGenomes)
basePop <- createVirginQueens(founderGenomes)
drones <- createDrones(basePop[1], nInd = 100)

# Mate a single virgin queen with 0 fathers
virginQueen <- basePop[2]
isVirginQueen(virginQueen)
(queen <- cross(basePop[2], fathers = drones[0]))
isVirginQueen(virginQueen)
queen # Not created

# Mate a Pop of virgin queens with 0 fathers
fatherGroups <- pullDroneGroupsFromDCA(drones, n = 2, nFathers = 0)
virginQueens <- basePop[3:5]
all(isVirginQueen(virginQueens))
(queens <- cross(virginQueens, fathers = fatherGroups[1:3]))
all(isVirginQueen(virginQueens))
queens # Not created

# Mate a Pop of virgin queens, but only some of them with 0 fathers
fatherGroups <- c(pullDroneGroupsFromDCA(drones, n = 2, nFathers = 0),
                  pullDroneGroupsFromDCA(drones, n = 2, nFathers = 10))
virginQueens <- basePop[6:9]
all(isVirginQueen(virginQueens))
(queens <- cross(virginQueens, fathers = fatherGroups))
isVirginQueen(virginQueens)
queens # Not created

# The other way around
fatherGroups <- fatherGroups[c(3,4,1,2)]
all(isVirginQueen(virginQueens))
(queens <- cross(virginQueens, fathers = fatherGroups))
isVirginQueen(virginQueens) # Stops at the first 0 fathers!!!
all(isDrone(drones))

# Mate a single colony to 0 fathers
colony <- createColony(basePop[16])
isVirginQueen(getVirginQueens(colony))
(colony <- cross(colony, fathers = fatherGroups[[4]]))
isVirginQueen(getVirginQueens(colony))

# Mate an apiary with 0 fathers
drones <- createDrones(basePop[1], nInd = 100)
fatherGroups <- pullDroneGroupsFromDCA(drones, n = 6, nFathers = 0)
apiary <- createColonies(basePop[10:15], n = 6)
all(isVirginQueen(mergePops(getVirginQueens(apiary))))
(apiary <- cross(apiary, fathers = fatherGroups))
all(isVirginQueen(mergePops(getVirginQueens(apiary))))
all(isDrone(drones))

# Mate apiary, but only some with 0 fathers
fatherGroups <- c(pullDroneGroupsFromDCA(drones, n = 3, nFathers = 10),
                  pullDroneGroupsFromDCA(drones, n = 3, nFathers = 0))
all(isVirginQueen(mergePops(getVirginQueens(apiary))))
(apiary <- cross(apiary, fathers = fatherGroups))
isVirginQueen(mergePops(getVirginQueens(apiary)))
all(isDrone(drones))
```

**Comment by janaobsteter** [comment]
Another question. If `checkMating = "error"`, the function stops at the first failed maiting. But if `checkMating = "warning"`, do we want to return one warning for all of them (currently, each failed mating throws an error, which could be a lot of text)?

**Comment by gregorgorjanc** [comment, marked as answer]
I think the best behaviour would be to proceed when we are in the warning mode. Don’t worry about too much warnings just now. We can address that later. At any rate, this will be an “advanced” simulation - most simulations will have more than zero fathers so this is just an extreme. Does this make sense?
